procsim/biomass/level1_product_generator.py

```diff
@@ -40,47 +55,360 @@
+        frame_bounds = [slice_start + d * self._frame_grid_spacing for d in range(NUM_FRAMES_PER_SLICE + 1)]
+
+        # Map to frame instances.
+        frames = []
+        for fi in range(len(frame_bounds) - 1):
+            # Continue if the acquisition start occurs after the frame bounds, or the acquisition end before.
+            if acq_start > frame_bounds[fi + 1] or acq_end < frame_bounds[fi]:
+                continue
+            frames.append(Frame(
+                id=fi + first_frame_nr,
+                validity_start=frame_bounds[fi],
+                validity_stop=frame_bounds[fi + 1] + self._frame_overlap,
+                sensing_start=max(frame_bounds[fi], acq_start),
+                sensing_stop=min(frame_bounds[fi + 1] + self._frame_overlap, acq_end),
+                status=FrameStatus.NOMINAL
+            ))
+
+        # If the first or last frame are too short, merge them with their neghbours.
+        if len(frames) > 1:
+            if frames[0].sensing_stop - frames[0].sensing_start < self._frame_lower_bound:
+                frames[1].sensing_start = frames[0].sensing_start
+                frames.pop(0)
+            if frames[-1].sensing_stop - frames[-1].sensing_start < self._frame_lower_bound:
+                frames[-2].sensing_stop = frames[-1].sensing_stop
+                frames.pop()
+
+        # If any frames are partial or merged, mark them as such.
+        for frame in frames:
+            if frame.sensing_stop - frame.sensing_start < frame.validity_stop - frame.validity_start:
+                frame.status = FrameStatus.PARTIAL
+            elif frame.sensing_stop - frame.sensing_start > frame.validity_stop - frame.validity_start:
+                frame.status = FrameStatus.MERGED
+
+        return frames
+
+    def _generate_product(self) -> None:
+        '''
+        Construct and write an output file given the variables set in this class.
+        '''
+        # Create a name generator, give it the right values and generate the product name.
+        name_gen = self._create_name_generator(self._hdr)
+        if self._creation_date is None:
+            self._creation_date = datetime.datetime.now(tz=datetime.timezone.utc)
+        name_gen.set_creation_date(self._creation_date)
+        name_gen._file_class = self._file_class
+        file_name = name_gen.generate_path_name()
+        os.makedirs(self._output_path, exist_ok=True)
+        full_file_name = os.path.join(self._output_path, file_name)
+
+        # Ensure the presence of vital variables.
+        if self._hdr.validity_start is None or self._hdr.validity_stop is None:
+            raise GeneratorError('Validity start/stop times must be known here.')
+        if self._hdr.acquisitions[0].slice_frame_nr is None:
+            raise GeneratorError('Frame number must be known here.')
+
+        # Construct the contents of the XML file.
+        root = et.Element('Earth_Explorer_File')
```

Comment:
This method is too long. 
Move the XML code to a separate method, or maybe even its own class (which allows for easier testing)

Reply:
I've moved it to a separate method for now.